### PR TITLE
[BUGFIX] - Validate regex to avoid unexpected exception

### PR DESCRIPTION
**Pull request by ivandrofly**
To see to problem add:
``` xml
  <RegularExpressions>
    <RegEx find="(" replaceWith="I" />
  </RegularExpressions>
```

Filename: **eng_OCRFixReplaceList.xml**

### CODE REVIEW

File: libse/Dictionaries/OcrFixReplaceList.cs [hunk 1]
```diff
@@ -113,14 +113,19 @@ public static OcrFixReplaceList FromLanguageId(string languageId)
...
                         {
                             string to = item.Attributes["replaceWith"].InnerText;
                             string from = item.Attributes["find"].InnerText;
-                            if (!list.ContainsKey(from))
-                                list.Add(from, to);
+                            if (Utilities.IsValidRegex(to) && Utilities.IsValidRegex(from))
+                            {
+                                if (!list.ContainsKey(from))
+                                    list.Add(from, to);
+                            }
```
Comment (xylographe) [inline libse/Dictionaries/OcrFixReplaceList.cs]: `to` contains a replacement pattern (see [Substitution Elements and Replacement Patterns](https://msdn.microsoft.com/en-us/library/ewy2t5e0(v=vs.110).aspx)).

Comment (xylographe) [inline libse/Dictionaries/OcrFixReplaceList.cs]: Also, **from** cannot be zero Length, but for **to** that's OK.

Comment (ivandrofly) [inline libse/Dictionaries/OcrFixReplaceList.cs]: :+1:  @xylographe

**Comment by ivandrofly** [comment]
:+1:  @xylographe, what do you mean from **to** its ok?

**Comment by xylographe** [comment]
The **to** string can be empty (Length == 0). It might seem weird, but if the regex pattern in **from** contains look-ahead and/or look-behind sub-expressions, it could happen.

An empty **from** string OTOH *must* be a mistake. For example:
```
PS> $re = [regex][string]::Empty
PS> $re.Replace('abc', 'x')
xaxbxcx
```
The empty string as a regex pattern matches everywhere.

**Comment by ivandrofly** [comment]
Nice :)

**Comment by ivandrofly** [comment]
:+1: /cc @niksedk

File: libse/Dictionaries/OcrFixReplaceList.cs [hunk 2]
```diff
@@ -113,14 +113,19 @@ public static OcrFixReplaceList FromLanguageId(string languageId)
...
                     {
-                        if (item.Attributes != null && item.Attributes["replaceWith"] != null && item.Attributes["find"] != null)
+                        if (item.Attributes != null && item.Attributes[ReplaceWith] != null && item.Attributes[Find] != null)
                         {
-                            string to = item.Attributes["replaceWith"].InnerText;
-                            string from = item.Attributes["find"].InnerText;
-                            if (!list.ContainsKey(from))
-                                list.Add(from, to);
+                            string to = item.Attributes[ReplaceWith].InnerText;
+                            string from = item.Attributes[Find].InnerText;
```
Comment (xylographe) [inline libse/Dictionaries/OcrFixReplaceList.cs]: Is this OK?
According to [MSDN](https://msdn.microsoft.com/en-us/library/system.xml.xmlattribute.innertext%28v=vs.110%29.aspx) the `XmlAttribute.InnerText` property would be **set** only.
Perhaps, better use the [`XmlAttribute.Value` property](https://msdn.microsoft.com/en-us/library/system.xml.xmlattribute.value%28v=vs.110%29.aspx) ?

Comment (ivandrofly) [inline libse/Dictionaries/OcrFixReplaceList.cs]: > According to MSDN the XmlAttribute.InnerText property would be set only.
> Perhaps, better use the XmlAttribute.Value property ?

:+1:

Comment (ivandrofly) [inline libse/Dictionaries/OcrFixReplaceList.cs]: I'm learning **PowerShell** any book you recommend? /cc @xylographe  :smile:

**Comment by ivandrofly** [comment]
I `force push` wrong branch :angry:

File: libse/Dictionaries/OcrFixReplaceList.cs (hunk 2 above)
Comment (xylographe) [inline libse/Dictionaries/OcrFixReplaceList.cs]: [Windows PowerShell in Action, Second Edition](https://www.manning.com/books/windows-powershell-in-action-second-edition)

It's about V2, hence, it doesn't address V3/V4 additions, but it's the best posh reference I know. It starts every subject simple (no previous knowledge required), before going into more complex details, until everything is revealed. :smile: 

It's miles ahead of other posh books I read, only downside is that there have been no upgrades for V3 and V4. Of course, MSDN contains info on changes between V2-V3, and V3-V4. And for the latest *usage* of any cmdlet, you should use `get-help -online` anyway.

Comment (ivandrofly) [inline libse/Dictionaries/OcrFixReplaceList.cs]: Nice, :smile: thanks ;)

**Pull request closed without merging**